DDG4/edm4hep/EDM4hepFileReader.cpp

```diff
@@ -0,0 +1,320 @@
+    }
+    auto const& floatKeys = source.template getKeys<float>();
+    for(auto const& key: floatKeys) {
+      m_fltValues[key] = source.template get<std::vector<float>>(key).value();
+    }
+    auto const& doubleKeys = source.template getKeys<double>();
+    for(auto const& key: doubleKeys) {
+      m_dblValues[key] = source.template get<std::vector<double>>(key).value();
+    }
+    using std::string;
+    auto const& stringKeys = source.template getKeys<string>();
+    for(auto const& key: stringKeys) {
+      m_strValues[key] = source.template get<std::vector<string>>(key).value();
+    }
+  }
+
+  /// get the parameters from the GenericParameters of the input EDM4hep run frame and store them in the RunParameters
+  /// extension
+  template <class T=podio::GenericParameters> void RunParameters::ingestParameters(T const& source) {
+    auto const& intKeys = source.template getKeys<int>();
+    for(auto const& key: intKeys) {
+      m_intValues[key] = source.template get<std::vector<int>>(key).value();
+    }
+    auto const& floatKeys = source.template getKeys<float>();
+    for(auto const& key: floatKeys) {
+      m_fltValues[key] = source.template get<std::vector<float>>(key).value();
+    }
+    auto const& doubleKeys = source.template getKeys<double>();
+    for(auto const& key: doubleKeys) {
+      m_dblValues[key] = source.template get<std::vector<double>>(key).value();
+    }
+    using std::string;
+    auto const& stringKeys = source.template getKeys<string>();
+    for(auto const& key: stringKeys) {
+      m_strValues[key] = source.template get<std::vector<string>>(key).value();
+    }
+  }
+
+  /// Class to read EDM4hep files
+  /**
+   *  \version 1.0
+   *  \ingroup DD4HEP_SIMULATION
+   */
+  class EDM4hepFileReader : public Geant4EventReader {
+  protected:
+    /// Reference to reader object
+    podio::ROOTReader m_reader {};
+    /// Name of the MCParticle collection to read
+    std::string m_collectionName;
+    /// Name of the EventHeader collection to read
+    std::string m_eventHeaderCollectionName;
+
+  public:
+    /// Initializing constructor
+    EDM4hepFileReader(const std::string& nam);
+
+    /// Read parameters set for this action
+    virtual EventReaderStatus setParameters(std::map< std::string, std::string >& parameters);
+
+    /// Read an event and fill a vector of MCParticles.
+    virtual EventReaderStatus readParticles(int event_number, Vertices& vertices, std::vector<Particle*>& particles);
+    /// Call to register the run parameters
+    void registerRunParameters();
+
+  };
+
+  /// Initializing constructor
+  dd4hep::sim::EDM4hepFileReader::EDM4hepFileReader(const std::string& nam)
+    : Geant4EventReader(nam)
+    , m_collectionName("MCParticles")
+    , m_eventHeaderCollectionName("EventHeader")
+  {
+    printout(INFO,"EDM4hepFileReader","Created file reader. Try to open input %s",nam.c_str());
+    m_reader.openFile(nam);
```

Comment:
```suggestion
    m_reader = podio::makeReader(nam);
```
to complement using the generic reader interface.

Reply:
Today is your lucky day https://github.com/AIDASoft/podio/releases/tag/v01-02 ;)

Reply:
I implemented the switch to Reader, which compiles except for a linking issue, which I hope will be solved in the new podio version. I reverted the commit with the implementation I have and can re-apply it when podio was updated in dev3 and it really works.

Reply:
https://github.com/AIDASoft/DD4hep/blob/abb85cbacd3909ea54733f2defd8053eaee63d12/DDG4/CMakeLists.txt#L108

`podio::podioRootIO` -> `podio::podioIO` might fix the linking issue (the `Reader` is not part of the root I/O for podio).